user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserOperator.java

```diff
@@ -4,146 +4,244 @@
         this.scramCredentialsOperator = scramCredentialsOperator;
         this.quotasOperator = quotasOperator;
         this.aclOperations = aclOperations;
         this.config = config;
+
+        this.selector = (config.getLabels() == null || config.getLabels().toMap().isEmpty()) ? new LabelSelector() : new LabelSelector(null, config.getLabels().toMap());
         this.passwordGenerator = new PasswordGenerator(this.config.getScramPasswordLength());
     }
 
-    @Override
-    public Future<Set<NamespaceAndName>> allResourceNames(String namespace) {
-        return CompositeFuture.
-                join(super.allResourceNames(namespace),
-                        config.isAclsAdminApiSupported() ? aclOperations.getAllUsers() : Future.succeededFuture(Set.of()),
-                        quotasOperator.getAllUsers(),
-                        config.isKraftEnabled() ? Future.succeededFuture(List.of()) : scramCredentialsOperator.getAllUsers() // SCRAM-SHA authentication is currently not supported when KRaft is used
-                ).map(compositeFuture -> {
-                    Set<NamespaceAndName> names = compositeFuture.resultAt(0);
-                    names.addAll(toResourceRef(namespace, compositeFuture.resultAt(1)));
-                    names.addAll(toResourceRef(namespace, compositeFuture.resultAt(2)));
-                    names.addAll(toResourceRef(namespace, compositeFuture.resultAt(3)));
-                    return names;
+    /**
+     * Gets all usernames which should be reconciled. They are collected from the Kubernetes resources as well as from
+     * the Kafka itself (based on existing ACLs, Quotas or SCRAM-SHA credentials). Querying the users also from Kafka
+     * is important to ensure proper deletion.
+     *
+     * @param namespace     Namespace where to look for the users
+     *
+     * @return  Set with Users and their namespaces
+     */
+    public CompletionStage<Set<NamespaceAndName>> getAllUsers(String namespace) {
+        // Get all users from KafkaUser resources
+        CompletionStage<Set<String>> kafkaUsers = getAllKafkaUserUsernames(namespace);
+
+        // Get the quota users
+        CompletionStage<Set<String>> quotaUsers = quotasOperator.getAllUsers();
+
+        // Get the ACL users
+        CompletionStage<Set<String>> aclUsers;
+        if (config.isAclsAdminApiSupported())   {
+            aclUsers = aclOperations.getAllUsers();
+        } else {
+            aclUsers = CompletableFuture.completedFuture(Set.of());
+        }
+
+        // Get the SCRAM-SHA users
+        CompletionStage<List<String>> scramUsers;
+        if (!config.isKraftEnabled()) {
+            scramUsers = scramCredentialsOperator.getAllUsers();
+        } else {
+            scramUsers = CompletableFuture.completedFuture(List.of());
+        }
+
+        return CompletableFuture.allOf(kafkaUsers.toCompletableFuture(), quotaUsers.toCompletableFuture(), aclUsers.toCompletableFuture(), scramUsers.toCompletableFuture())
```

Comment:
The `CompletionStage` seems to be the higher interface so I think it is better to use it instead of using `CompletableFuture` directly. It does not have any significant reason. It just seems better to use one thing everywhere instead of mixing both of them depending on what comes handy today.

Reply:
I am totally with you on using interface to handle this kind of things. What I see in this case is that the `CompletableFuture` class implements both `Future` and `CompletionStage`. Dealing with a `CompletableFuture` instance allows to deal with the steps on async calls (so using `CompletionStage` methods) as well as allows to get the result of an async call (which is part of the `Future`, avoidint to call the `toCompletableFuture` every time).
That was the rationale behind my comment.

Reply:
TBH, I'm not entirely sure what you are suggesting now.

Reply:
That maybe we should use the CompletableFuture everywhere to get both the CompletionStage and Future implementation with no need to use toCompletableFuture method so often.
Anyway it could be a bigger change, so let's stick with current status for now.